integrations/access/opsgenie/app.go

```diff
@@ -17,15 +17,575 @@ limitations under the License.
+			DialOpts: []grpc.DialOption{
+				grpc.WithConnectParams(grpc.ConnectParams{Backoff: bk, MinConnectTimeout: initTimeout}),
+				grpc.WithReturnConnectionError(),
+			},
+		}); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	if _, err = a.checkTeleportVersion(ctx); err != nil {
+		return trace.Wrap(err)
+	}
+
+	a.opsgenie, err = NewClient(a.conf.ClientConfig)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+func (a *App) checkTeleportVersion(ctx context.Context) (proto.PingResponse, error) {
+	log := logger.Get(ctx)
+	log.Debug("Checking Teleport server version")
+
+	pong, err := a.teleport.Ping(ctx)
+	if err != nil {
+		if trace.IsNotImplemented(err) {
+			return pong, trace.Wrap(err, "server version must be at least %s", minServerVersion)
+		}
+		log.Error("Unable to get Teleport server version")
+		return pong, trace.Wrap(err)
+	}
+	err = lib.AssertServerVersion(pong, minServerVersion)
+	return pong, trace.Wrap(err)
+}
+
+func (a *App) onWatcherEvent(ctx context.Context, event types.Event) error {
+	if kind := event.Resource.GetKind(); kind != types.KindAccessRequest {
+		return trace.Errorf("unexpected kind %s", kind)
+	}
+	op := event.Type
+	reqID := event.Resource.GetName()
+	ctx, _ = logger.WithField(ctx, "request_id", reqID)
+
+	switch op {
+	case types.OpPut:
+		ctx, _ = logger.WithField(ctx, "request_op", "put")
+		req, ok := event.Resource.(types.AccessRequest)
+		if !ok {
+			return trace.Errorf("unexpected resource type %T", event.Resource)
+		}
+		ctx, log := logger.WithField(ctx, "request_state", req.GetState().String())
+
+		var err error
+		switch {
+		case req.GetState().IsPending():
+			err = a.onPendingRequest(ctx, req)
+		case req.GetState().IsApproved():
+			err = a.onResolvedRequest(ctx, req)
+		case req.GetState().IsDenied():
+			err = a.onResolvedRequest(ctx, req)
+		default:
+			log.WithField("event", event).Warn("Unknown request state")
+			return nil
+		}
+
+		if err != nil {
+			log.WithError(err).Error("Failed to process request")
+			return trace.Wrap(err)
+		}
+
+		return nil
+	case types.OpDelete:
+		ctx, log := logger.WithField(ctx, "request_op", "delete")
+
+		if err := a.onDeletedRequest(ctx, reqID); err != nil {
+			log.WithError(err).Error("Failed to process deleted request")
+			return trace.Wrap(err)
+		}
+		return nil
+	default:
+		return trace.BadParameter("unexpected event operation %s", op)
+	}
+}
+
+func (a *App) onPendingRequest(ctx context.Context, req types.AccessRequest) error {
+	if len(req.GetSystemAnnotations()) == 0 {
+		logger.Get(ctx).Debug("Cannot proceed further. Request is missing any annotations")
+		return nil
+	}
+
+	// First, try to create a notification alert.
+	isNew, notifyErr := a.tryNotifyService(ctx, req)
+	notifyErr = trace.Wrap(notifyErr)
+
+	// To minimize the count of auto-approval tries, lets attempt it only when we just created an alert.
+	// But if there's an error, we can't really know is the alert new or not so lets just try.
```

Comment:
```
// But if there's an error, we can't really know is the alert new or not so lets just try.
```
I'm not sure what you're saying here. Do you mean to say we can't differentiate between failing to create a notification because it already exists, vs other failure cases?

Reply:
Yeah, that's the intention.